From: Jeff Mahoney <jeffm@suse.com>
Subject: [PATCH] blockdev: fix race between mount/umount
Suse-bugzilla: 63178
Obsoletes: patches.fixes/bd_claim_flags.diff

This patch fixes a race between mount and umount, when set_blocksize may
reset the blocksize after the new filesystem is already mounted. This race
can manifest itself as hangs, buffer errors, and possibly other symptoms.

After trying two other methods, Linus suggested this as a solution. I've run
a 24-hour stress test on it and it has passed successfully.

This does meant the block size is not restored to the default value after the
final umount. However, userspace openers aren't affected by the block size
except for performance reasons. If the opener requirs a specific block size,
it should set it explicitly. Depending on a filesystem being mounted or not
is quite racy to begin with.

It would be possible to reset the blocksize on the final close if this is a
must-have, but I don't believe it to be.

Signed-off-by: Jeff Mahoney <jeffm@suse.com>

diff -ruNpX dontdiff linux-2.6.11/fs/super.c linux-2.6.11.bs/fs/super.c
--- linux-2.6.11/fs/super.c	2005-03-16 15:06:50.000000000 -0500
+++ linux-2.6.11.bs/fs/super.c	2005-03-15 16:24:50.000000000 -0500
@@ -732,7 +732,7 @@ void kill_block_super(struct super_block
 
 	bdev_uevent(bdev, KOBJ_UMOUNT);
 	generic_shutdown_super(sb);
-	set_blocksize(bdev, sb->s_old_blocksize);
+	sync_blockdev (bdev);
 	close_bdev_excl(bdev);
 }
 
